epitator/annospan.py

```diff
@@ -140,3 +134,37 @@ def iterate_leaf_base_spans(self):
         for span in self.iterate_base_spans():
             if not isinstance(span, SpanGroup):
                 yield span
+
+    def combined_metadata(self):
+        """
+        Return the merged metadata dictionaries from all descendant spans.
+        Presedence of matching properties follows the order of a pre-order tree traversal.
+        """
+        leaf_spans = list(self.iterate_base_spans())
+        leaf_spans.reverse()
+        result = {}
+        for leaf_span in leaf_spans + [self]:
+            if leaf_span.metadata:
+                result.update(leaf_span.metadata)
+        return result
+
+
+class SpanGroup(AnnoSpan):
+    """
+    A AnnoSpan that extends through a group of AnnoSpans.
+    """
+    def __init__(self, base_spans, label=None):
+        assert isinstance(base_spans, list)
+        assert len(base_spans) > 0
+        super(SpanGroup, self).__init__(
```

Comment:
I think it would be better to create new SpanGroups than dynamically modify their instantiation variables. Keeping data structures as static as possible generally leads to simpler code and prevents bugs.

Reply:
## `start` and `end`

I think that for AnnoSpans, `start` and `end` are sort of… part of the core definition. But for a SpanGroup, they have no intrinsic meaning — they're just labels — and I think it makes sense to reference them to the lowest start and highest end of the base_spans. Otherwise, if you, say, applied the function that's defined elsewhere here which trims base spans to one of the base spans, and it was part of a SpanGroup, it would be wrong.

## Creating new SpanGroups vs. modifying an existing SpanGroup

And the reason for that is that, if there was code elsewhere that was using that name to access the object, changing the object in place could cause unexpected behavior, and it's better to create new objects?

I still think that this is useful, but I'm thinking about a future use case where, for instance, we could use this sort of class to generate incidents. (I guess this is what's happening in the structured data annotator).

It seems to me that creating new SpanGroups each time you want to add a property to a SpanGroup would be fine, as long as you had a method for instantiating a SpanGroup by copying and modifying an existing SpanGroup? I don't think I'm making my argument clearly here.

Here's my question, in an example. Say you wanted to add a date to an infection span based on a set of rules. You identify the spans you want to combine, and do `group1 = SpanGroup([infectionspan, datespan])`.

Then you have code that identifies a GeoSpan that's nearby. If you do `group2 = SpanGroup([group1, geospan])`, then you have a weird nested structure, and so on, if you got to the point where you were resolving multiple properties in that way. You're probably talking about doing something like `group2 = SpanGroup([group1.base_spans, geospan])`, which works better.

Something like `group2 = copy(group1).add_span(geospan)` would both copy the object and add a span (this is imagining an `.add_span()` method which appended that AnnoSpan to the `base_span` list), but this again seems like a rather unsatisfying solution.

I'll show you how I solved this, and then we can discuss.